### PR TITLE
Implement QUIC abort exceptions, fix channel usage, and implement some tests.

### PR DESCRIPTION
**Pull request by scalablecory**
New: implement QuicStreamAbortedException, QuicConnectionAbortedException, QuicOperationAbortedException.
Fix: MsQuic returning null implementation when completing listener/connection channels.
Fix: some flags types being defined as 8-bit when they are 32-bit in msquic.h
Reorganize tests a little bit and avoid a shared listener creating variability.
Add some additional tests demonstrating failures seen in HttpClient. Issues created for those: https://github.com/dotnet/runtime/issues/32050 https://github.com/dotnet/runtime/issues/32049 https://github.com/dotnet/runtime/issues/32048

### CODE REVIEW

**Comment by jkotalik** [comment]
> Fix: MsQuic returning null implementation when completing listener/connection channels.

I was just about to do this 😄 

> Reorganize tests a little bit and avoid a shared listener creating variability.

I thought the listener was created once per test still. Did you see this variability?

**Comment by scalablecory** [comment]
some http3 code snuck into the commit; fixed now :)

**Comment by scalablecory** [comment]
> I thought the listener was created once per test still. Did you see this variability?

Yes, xunit test classes are newed up once per test so it wasn't like a race condition. More for organization purposes (I'm worried about unintentionally using it in a way that exercises race conditions VS a specific test -- this is more explicit and in line with how most of our other tests function)

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs
Comment (jkotalik) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs]: Make sendErrorCode a local variable. I don't believe it is used elsewhere.

File: src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs
Comment (jkotalik) [inline src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs]: I think this is more of a style thing. I personally prefer initialization of the listener in the constructor, but I'm not strongly opposed to making it explicit

**Review by jkotalik — APPROVED** [review]

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs
Comment (scalablecory) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs]: It is used to throw a `QuicStreamAbortedException` on the other side of the `ValueTask`

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs
Comment (halter73) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs]: I think it's better to continue returning null than to throw an exception from AcceptStreamAsync when the connection is closing. The connection being closed is a common enough that throwing an exception feels noisy. I know this will make the QuicConnection.AcceptStreamAsync() wrapper a little more complicated, but I think it's worth it.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs
Comment (halter73) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs]: I think we should continue returning null here too.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs
Comment (jkotalik) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs]: Yeah, looking back on this, I agree as well.

File: src/libraries/System.Net.Quic/ref/System.Net.Quic.cs
Comment (halter73) [inline src/libraries/System.Net.Quic/ref/System.Net.Quic.cs]: I see that QuicException derives directly from Exception. Would it be better to have it derive from IOException or possibly OperationCanceledException if it's always user-initiated?

I would lean towards IOException given how it's currently used. I see it's being triggered by stuff like "HandleEventShutdownInitiatedByPeer" which feels really IOExceptiony to me.

Comment (scalablecory) [reply src/libraries/System.Net.Quic/ref/System.Net.Quic.cs]: Yea that seems reasonable. Open up an issue for it?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs
Comment (scalablecory) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs]: I think that anyone familiar with sockets will expect consistency with that, which is why I went that direction, but I also think most everyone else might expect it will return null. The question becomes who is the more likely user. It's trouble to merge Stream and Socket 🙃.

I don't feel strongly about it. We can change it.

Comment (jkotalik) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs]: Let's keep it with the "current" behavior of returning null, but fix it in QuicConnection.AcceptStreamAsync.

File: src/libraries/System.Net.Quic/ref/System.Net.Quic.cs
Comment (jkotalik) [reply src/libraries/System.Net.Quic/ref/System.Net.Quic.cs]: I also don't know if QuicOperationAbortedException will be necessary now if we start returning null.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs
Comment (scalablecory) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs]: Actually, I think this needs some more thought now. We should consider how consistent the APIs between QuicStream, QuicConnection, QuicListener need to be in how they behave on disposal. QuicStream.ReadAsync() etc. do not return a nullable type and will need to throw an exception.

Comment (scalablecory) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs]: I'm going to commit this for now to unblock things (the APIs are currently broken, returning null triggers a bug) but will open an issue to figure this out.

Comment (scalablecory) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs]: Issue: https://github.com/dotnet/runtime/issues/32142

**Comment by maryamariyan** [comment]
Note regarding the `new-api-needs-documentation` label:
This serves as a reminder for when your PR is modifying a ref *.cs file and adding/modifying public APIs, to please make sure the API implementation in the src *.cs file is documented with triple slash comments, so the PR reviewers can sign off that change.

**Comment by karelz** [comment]
@maryamariyan this will be an interesting case as the API is in new library that we intentionally don't ship yet. The API has to go through API iterations and then API review. Then it will be good time to document it once we have final shape.
How does that fit our API documentation flow?

File: src/libraries/System.Net.Quic/ref/System.Net.Quic.cs
Comment (halter73) [reply src/libraries/System.Net.Quic/ref/System.Net.Quic.cs]: I looked at SocketException, and it doesn't derive from IOException, so I'm no longer convinced this is the way to go. Maybe it should derive from SystemException (which SocketException does indirectly), but I'm not sure what the guidance there is. Is this supposed to be the base class for all Exceptions in a System.* namespace?

**Pull request merged**
